src/trace/layers/ordered_leaf.rs

```diff
@@ -0,0 +1,415 @@
+where
+    K: Ord + Clone + Display,
+    R: Eq + HasZero + AddAssignByRef + Clone + Display,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        let TrieSlice(storage, cursor) = self;
+        let mut cursor: OrderedLeafCursor = cursor.clone();
+
+        while cursor.valid(storage) {
+            let (key, val) = cursor.key(storage);
+            writeln!(f, "{} -> {}", key, val)?;
+            cursor.step(storage);
+        }
+
+        Ok(())
+    }
+}
+
+// TODO: by-value merge
+impl<K, R> Add<Self> for OrderedLeaf<K, R>
+where
+    K: Ord + Clone,
+    R: Eq + HasZero + AddAssignByRef + Clone,
+{
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        if self.keys() == 0 {
+            rhs
+        } else if rhs.keys() == 0 {
+            self
+        } else {
+            self.merge(&rhs)
+        }
+    }
+}
+
+impl<K, R> AddAssign<Self> for OrderedLeaf<K, R>
+where
+    K: Ord + Clone,
+    R: Eq + HasZero + AddAssignByRef + Clone,
+{
+    fn add_assign(&mut self, rhs: Self) {
+        if rhs.keys() != 0 {
+            *self = self.merge(&rhs);
+        }
+    }
+}
+
+impl<K, R> AddAssignByRef for OrderedLeaf<K, R>
+where
+    K: Ord + Clone,
+    R: Eq + HasZero + AddAssignByRef + Clone,
+{
+    fn add_assign_by_ref(&mut self, other: &Self) {
+        if other.keys() != 0 {
+            *self = self.merge(other);
+        }
+    }
+}
+
+impl<K, R> AddByRef for OrderedLeaf<K, R>
+where
+    K: Ord + Clone,
+    R: Eq + HasZero + AddAssignByRef + Clone,
+{
+    fn add_by_ref(&self, rhs: &Self) -> Self {
+        self.merge(rhs)
+    }
+}
+
+impl<K, R> NegByRef for OrderedLeaf<K, R>
+where
+    K: Ord + Clone,
+    R: NegByRef,
+{
+    fn neg_by_ref(&self) -> Self {
+        Self {
+            vals: self
+                .vals
+                .iter()
+                .map(|(k, v)| (k.clone(), v.neg_by_ref()))
+                .collect(),
+        }
+    }
+}
+
+impl<K, R> Neg for OrderedLeaf<K, R>
+where
+    K: Ord + Clone,
+    R: Neg<Output = R>,
+{
+    type Output = Self;
+
+    fn neg(self) -> Self {
+        Self {
+            vals: self.vals.into_iter().map(|(k, v)| (k, v.neg())).collect(),
+        }
+    }
+}
+
+impl<K, R> NumEntries for OrderedLeaf<K, R>
+where
+    K: Ord + Clone,
+    R: Eq + HasZero + AddAssignByRef + Clone,
+{
+    fn num_entries_shallow(&self) -> usize {
+        self.keys()
+    }
+
+    fn num_entries_deep(&self) -> usize {
+        self.keys()
+    }
+
+    fn const_num_entries() -> Option<usize> {
+        None
+    }
+}
+
+impl<K, R> SharedRef for OrderedLeaf<K, R>
+where
+    K: Clone,
+    R: Clone,
+{
+    type Target = Self;
+
+    fn try_into_owned(self) -> Result<Self::Target, Self> {
+        Ok(self)
+    }
+}
+
+/// A builder for unordered values.
+pub struct OrderedLeafBuilder<K, R> {
+    /// Unordered values.
+    pub vals: Vec<(K, R)>,
+}
+
+impl<K: Ord + Clone, R: Eq + HasZero + AddAssignByRef + Clone> Builder
```

Comment:
Maybe move all of these trait bounds to a where clause, it's kinda hard to get the gist of the impl block when you have to figure out where the traits end and the types begin